File: Problems/0211. (M) Design Add and Search Words Data Structure/Python3/solution.py
class WordDictionary:
  # prefix tree (trie) that supports wildcard search

  class Node():
    def __init__(self, last=False):
      self.children = [None] * len(string.ascii_lowercase)
      self.last = last

  def __init__(self):
    self.root = WordDictionary.Node()
    # delete unsupported, keep track of longest word in tree
    self.max_length = 0

  def addWord(self, word: str) -> None:
    # iteratively add word to trie

    cur = self.root
    for c in word:
      idx = ord(c) - ord('a')
      if cur.children[idx] is None:
        cur.children[idx] = WordDictionary.Node()
      cur = cur.children[idx]
    cur.last = True
    self.max_length = max(self.max_length, len(word))

  def search(self, word: str, node: Optional[Node] = None) -> bool:
    # sanity check
    if len(word) > self.max_length:
      return False

    if not node:
      node = self.root
    
    for i, c in enumerate(word):
      if c == '.':
        ret = False
        for descendant in node.children:
          # recurse on wildcards
          if descendant:
            ret |= self.search(word[i+1:], descendant)
        return ret

      idx = ord(c) - ord('a')
      if not node.children[idx]:
        return False
      node = node.children[idx]
    return node.last

File: Problems/0211. (M) Design Add and Search Words Data Structure/Java/solution.java
class WordDictionary {
  // prefix tree (trie) that supports wildcard search
  Node root;
  int max_length;

  class Node {
    Node[] children;
    boolean last;

    public Node() {
      children = new Node[26];
      last = false;
    }
  }

  public WordDictionary() {
    root = this.new Node();
    // since WordDictionary doesn't support delete
    // keep track of length of longest word
    max_length = 0;
  }
  
  public void addWord(String word) {
    // iteratively add word to trie

    Node cur = this.root;
    int idx;
    for (int i = 0; i < word.length(); i++) {
      idx = (int) word.charAt(i) - (int) 'a';
      if (cur.children[idx] == null)
        cur.children[idx] = this.new Node();
      cur = cur.children[idx];
    }
    cur.last = true;
    this.max_length = Math.max(this.max_length, word.length());
  }
  
  public boolean search(String word, Node... node) {
    // sanity check
    if (word.length() > this.max_length)
      return false;

    Node cur = this.root;
    // some weirdness going on with Varargs
    // apparently Node... is just syntactic sugar for Node[]
    if (node.length == 1)
      cur = node[0];

    int idx = 0;
    for (int i = 0; i < word.length(); i++) {
      if (word.charAt(i) == '.') {
        boolean ret = false;
        for (int j = 0; j < 26; j++) {
          // recurse on wildcards
          if (cur.children[j] != null)
            ret |= search(word.substring(i+1), cur.children[j]);
        }
        return ret;
      }
      idx = (int) word.charAt(i) - (int) 'a';
      if (cur.children[idx] == null)
        return false;
      cur = cur.children[idx];
    }
    return cur.last;
  }
}
